### PR TITLE
Traits cleanup

### PR DESCRIPTION
**Pull request by DerManoMann**
# WIP #

Some cleanup and fixes for trait handling (as it was intended IMHO).

* Deal with nested traits (traits using other traits)
* Fix operationId for traits and interfaces
* Fix some quirk in `fullyQualifiedName()` that would modify itself and break things further down the track
* Use correct var for trait property names in `ImportTraits` processor
* Added traits example as reference

The result of these changes is that **all** trait properties are merged into the class using the trait. A side effect is that the traits are still showing up as schema in the output, but are basically empty.

I suppose the alternative would be to **not** merge properties and use `allOf` on classes that use traits. Not sure yet what is better (or more `correct`), though.

Some discussion would be useful before processing any further....

### CODE REVIEW

**Comment by mroeling** [comment]
Great effort! I think it fits the expected result for #750.

Looking forward to see the use of references as well. Something like: Product extends SimpleProduct, with extra traits.

**Comment by DerManoMann** [comment]
> Looking forward to see the use of references as well. Something like: Product extends SimpleProduct, with extra traits.

Hmm, I would have thought inheriting from parent classes should work as. If you referring to the use of `allOf`, that is an alternative approach but not something that would happen at the same time as merging properties?

**Comment by DerManoMann** [comment]
So, a few  more questions to discuss...
1. Currently the validation will log a warning due to a [duplicate property](https://github.com/DerManoMann/swagger-php/blob/traits-example/Examples/using-traits/Product.php#L24).
Technically, this is valid and the class property/annotation should win (and it does). So, should the warning be supressed if the property comes from a trait? Maybe I am just not sure what to do with those warnings...  
1. Looking at the output the traits are all empty (merged) and look funny in swagger-ui. Should they be removed as nothing is referencing them either?
1. I might add a separate processor for importing traits as `allOf` refs instead of merging properties. Maybe people could then chose one or the other?

**Comment by mroeling** [comment]
In general: IMHO no functionality should be removed when parsing @OA into (yml) output.

1) Although it might be possible to double define a property, IMHO it's more like bad design... So I would prefer to have the warning in these cases. Since it is possible it doesn't have to be a failure, but a warning should fit the situation.

2) and 3) OA does support references, both for the API calls and the allOf implementation. So that's one of the reasons to prefer to have a (separate) processor. Another reason is rather personal. Our company not only uses the generated YML (or json, since we use 2.x because of the lack of support yet in 3.x) for the API calls, but also as base for creating models and interfaces in Typescript, using the swagger-schema-official and swagger-parser packages.
From the OA documentation: 
> An OpenAPI definition can then be used by documentation generation tools to display the API, **code generation tools to generate servers and clients** in various programming languages, testing tools, and many other use cases.
This would not be possible when traits are deleted.

**Comment by bfanger** [comment]
@zircote Could you add @DerManoMann as collaborator?

File: tests/ExamplesTest.php [hunk 1]
```diff
@@ -31,15 +36,16 @@ public function testExample($example, $output)
...
-            ['misc', 'misc.json'],
-            ['openapi-spec', 'openapi-spec.json'],
-            ['petstore.swagger.io', 'petstore.swagger.io.json'],
-            ['petstore-3.0', 'petstore-3.0.json'],
-            ['swagger-spec/petstore', 'petstore.json'],
-            ['swagger-spec/petstore-simple', 'petstore-simple.json'],
-            ['swagger-spec/petstore-with-external-docs', 'petstore-with-external-docs.json'],
-            ['using-refs', 'using-refs.json'],
-            ['example-object', 'example-object.json'],
+//            ['misc', 'misc.json', []],
```
Comment (bfanger) [inline tests/ExamplesTest.php:39]: Is it possible to add a custom command line option to phpunit? Something like:
`phpunit --swagger-example using-traits` 
Right now the other examples are excluded from running in travisCI and easy to forget to re-enable when te PR is ready.

Comment (DerManoMann) [reply tests/ExamplesTest.php:39]: Yeah, that was an oversight from doing manual testing. I'll restore that once we are happy with the changes.
I guess I could user `--filter using-traits` but then it gets complicated when trying to debug in the IDE...

**Comment by DerManoMann** [comment]
@mroeling thanks for the comments. 

I think I agree with you on all points and unless there is more opposition I'll continue working on a new processor that merges traits as `allOf` refs.

Also, I think we just have to separate the annotation processing from the PHP level rules about conflicts and overriding of properties, etc.

That means it totally makes sense to keep all the validation warnings in place and its the authors responsibility to keep things sane :)

**Comment by mroeling** [comment]
@DerManoMann Great!

**Comment by DerManoMann** [comment]
Closing this as it was never intended to be merged in this form. All relevant code changes are covered by other PRs.

**Pull request closed without merging**